### PR TITLE
feat: Add template-admin + user-admin role for managing templates + users

### LINKED ISSUE #2135: Create lesser admin role
## Problem statement

Our admin user has full shell access to every workspace, letting them steal secrets and manipulate workspaces at will.  We have clear instructions from our customers that this is a Big Bad. Our major deployments will disable the existing admin users and will want a lesser admin role.

## Definition of done

There is a lesser role than the current `admin` role, without the ability to enter the workspace's terminals. The current admin role is renamed to something more specific.

## Suggestions

I propose we rename "admin" to "root" or "deployer". I prefer "root" since everyone knows you're not supposed to use "root", except transiently. Everyone also knows "root" is dangerous.

There is some precedent for this pattern in GitHub Enterprise, which has a special deployer access panel that is not associated with any user.

cc @Emyrk

_Originally posted by @ammario in https://github.com/coder/coder/issues/2108#issuecomment-1149085452_

### PR DESCRIPTION
**Pull request by Emyrk**
# What this does

Adds a `template-admin` role that has full CRUD to templates and files. 
Adds a `user-admin` role that has full CRUD on users

Thoughts? Trying to partially solve https://github.com/coder/coder/issues/2135 by removing the need to promote everyone to admin.

Screenshot from 2022-08-12 16-02-28

# Bonus

UI only shows assignable roles. So if you cannot assign a role, it won't appear in the dropdown. (Except for self assigning roles)

# TODO:

Next PR will migrate "Admin" -> "Template Admin, User Admin"
And migrate the first user from "Admin" -> "Owner"

### CODE REVIEW

File: coderd/roles.go [hunk 1]
```diff
@@ -13,31 +13,44 @@ import (
 
 // assignableSiteRoles returns all site wide roles that can be assigned.
 func (api *API) assignableSiteRoles(rw http.ResponseWriter, r *http.Request) {
-	// TODO: @emyrk in the future, allow granular subsets of roles to be returned based on the
-	// 	role of the user.
-
+	actorRoles := httpmw.AuthorizationUserRoles(r)
```
Comment (mafredri) [inline coderd/roles.go:16]: Any specific reason we're doing this before `api.Authorize`?

File: coderd/roles.go [hunk 2]
```diff
@@ -13,31 +13,44 @@ import (
...
+		}
+	}
+
+	httpapi.Write(rw, http.StatusOK, convertRoles(assignable))
 }
 
 // assignableSiteRoles returns all site wide roles that can be assigned.
 func (api *API) assignableOrgRoles(rw http.ResponseWriter, r *http.Request) {
 	// TODO: @emyrk in the future, allow granular subsets of roles to be returned based on the
 	// 	role of the user.
```
Comment (mafredri) [inline coderd/roles.go:36]: Is this TODO TODOne, like the previous one?

**Review by mafredri — APPROVED** [review]
This looks good to me (although, do consider that my RBAC lingo is weak).

I wonder how much this will help towards #2135, though. I imagine #2135 is looking for a user that can also delete workspaces, but not inspect them.

File: coderd/roles.go (hunk 1 above)
Comment (Emyrk) [reply coderd/roles.go:16]: It is just pulling the roles from the actor from the middleware. This doesn't execute any code. Just pulling state from the request.

**Comment by Emyrk** [comment]
@mafredri I made a `user-admin` role as well. Combined, I think they cover what an "admin" wants to be, and blocks workspace execution on other user's workspaces

File: coderd/rbac/builtin_test.go [hunk 3]
```diff
@@ -165,44 +169,54 @@ func TestRolePermissions(t *testing.T) {
...
+			},
+		},
+		{
+			Name: "MyWorkspaceInOrgExecution",
+			// When creating the WithID won't be set, but it does not change the result.
+			Actions:  []rbac.Action{rbac.ActionCreate, rbac.ActionRead, rbac.ActionUpdate, rbac.ActionDelete},
+			Resource: rbac.ResourceWorkspaceExecution.InOrg(orgID).WithOwner(currentUser.String()),
+			AuthorizeMap: map[bool][]authSubject{
+				true:  {admin, orgAdmin, orgMemberMe},
+				false: {memberMe, otherOrgAdmin, otherOrgMember, templateAdmin, userAdmin},
```
Comment (Emyrk) [inline coderd/rbac/builtin_test.go:183]: Workspace execution can only be done by `admin`, `orgadmin` and the owner of the workspace.

File: coderd/rbac/builtin_test.go [hunk 4]
```diff
@@ -124,9 +125,12 @@ func TestRolePermissions(t *testing.T) {
 	otherOrgMember := authSubject{Name: "org_member_other", UserID: uuid.NewString(), Roles: []string{rbac.RoleMember(), rbac.RoleOrgMember(otherOrg)}}
 	otherOrgAdmin := authSubject{Name: "org_admin_other", UserID: uuid.NewString(), Roles: []string{rbac.RoleMember(), rbac.RoleOrgMember(otherOrg), rbac.RoleOrgAdmin(otherOrg)}}
 
+	templateAdmin := authSubject{Name: "template-admin", UserID: templateAdminID.String(), Roles: []string{rbac.RoleMember(), rbac.RoleTemplateAdmin()}}
+	userAdmin := authSubject{Name: "template-admin", UserID: templateAdminID.String(), Roles: []string{rbac.RoleMember(), rbac.RoleUserAdmin()}}
```
Comment (mafredri) [inline coderd/rbac/builtin_test.go:129]:
```suggestion
	userAdmin := authSubject{Name: "user-admin", UserID: templateAdminID.String(), Roles: []string{rbac.RoleMember(), rbac.RoleUserAdmin()}}
```

File: coderd/workspaceagents.go [hunk 5]
```diff
@@ -70,7 +70,7 @@ func (api *API) workspaceAgentDial(rw http.ResponseWriter, r *http.Request) {
 
 	workspaceAgent := httpmw.WorkspaceAgentParam(r)
 	workspace := httpmw.WorkspaceParam(r)
-	if !api.Authorize(r, rbac.ActionUpdate, workspace) {
+	if !api.Authorize(r, rbac.ActionCreate, workspace.ExecutionRBAC()) {
```
Comment (mafredri) [inline coderd/workspaceagents.go:73]: I didn't quite catch why this changed from Update to Create. My intuition says it doesn't really matter. So it's probably fine but just checking.

**Review by mafredri — APPROVED** [review]
The new role was a nice addition!

File: coderd/workspaceagents.go (hunk 5 above)
Comment (Emyrk) [reply coderd/workspaceagents.go:73]: It doesn't. It's arbitrary.

But what I did change was I added `ResourceWorkspaceExecution` as a unique resource. So giving CRUD to a workspace **does not** grant execution via a tunnel. These are now distinct permissions.

**Review by Kira-Pilot — APPROVED** [review]
nicely done

**Pull request merged**